File: healer/src/weakhandle.rs
// SPDX-License-Identifier: GPL-2.0-or-later
/*
 * Copyright (C) 2025 Oracle.  All Rights Reserved.
 * Author: Darrick J. Wong <djwong@kernel.org>
 */
use crate::baddata;
use crate::badness;
use crate::getmntent::MountEntries;
use crate::xfs_fs::xfs_fid;
use crate::xfs_fs::xfs_fsop_geom;
use crate::xfs_fs::xfs_fsop_handlereq;
use crate::xfs_fs::xfs_handle;
use crate::xfs_types::XfsFid;
use crate::xfsprogs::M_;
use nix::ioctl_readwrite;
use nix::libc::O_LARGEFILE;
use std::ffi::OsString;
use std::fmt::Display;
use std::fmt::Formatter;
use std::fs::File;
use std::io::Error;
use std::io::ErrorKind;
use std::io::Result;
use std::os::fd::AsRawFd;
use std::os::raw::c_void;
use std::os::unix::ffi::OsStringExt;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::Arc;

ioctl_readwrite!(xfs_ioc_fd_to_handle, 'X', 106, xfs_fsop_handlereq);

/* just pick a value we know is more than big enough */
const MAXHANSIZ: usize = 64;

impl PartialEq for xfs_handle {
    fn eq(&self, other: &Self) -> bool {
        // SAFETY: accessing an arm of a union that exists only to force memory alignment
        unsafe { self.ha_u._ha_fsid == other.ha_u._ha_fsid && self.ha_fid == other.ha_fid }
    }
}

impl TryFrom<&File> for xfs_handle {
    type Error = Error;

    /// Create an xfs_handle for an open file
    fn try_from(fp: &File) -> Result<xfs_handle> {
        assert!(MAXHANSIZ >= std::mem::size_of::<xfs_handle>());

        let mut value: Vec<u8> = vec![0; MAXHANSIZ];
        let mut hreq: xfs_fsop_handlereq = Default::default();
        let mut hlen: u32 = 0;

        hreq.fd = fp.as_raw_fd() as u32;
        hreq.oflags = O_LARGEFILE as u32;
        hreq.ohandle = value.as_mut_ptr() as *mut c_void;
        hreq.ohandlen = &mut hlen;

        // SAFETY: Trusting the kernel not to corrupt hreq, value, or anything else.  This is wildly
        // incorrect because the kernel interface does not require userspace to pass in the size of
        // the object ohandle, so it writes blindly to *ohandle.
        unsafe {
            xfs_ioc_fd_to_handle(fp.as_raw_fd(), &mut hreq)?;
        }
        if hlen as usize != std::mem::size_of::<xfs_handle>() {
            return Err(baddata!(M_("Bad file handle size"), xfs_handle, hlen));
        }

        // SAFETY: We asserted above that value is large enough to store an xfs_handle, so we can
        // cast and struct copy here.
        unsafe {
            let hanp: *const xfs_handle = value.as_ptr() as *const xfs_handle;
            let ret: xfs_handle = *hanp;
            Ok(ret)
        }
    }
}

/// Filesystem handle that can be disconnected from any open files
pub struct WeakHandle {
    /// device for the xfs filesystem
    fsname: String,

    /// path to the filesystem mountpoint
    mountpoint: Arc<PathBuf>,

    /// Filesystem handle
    handle: xfs_handle,

    /// Does this filesystem support parent pointers?
    has_parent: bool,
}

impl WeakHandle {
    /// Try to reopen the filesystem with a given mountpoint
    fn reopen_from(&self, mountpoint: &Path) -> Result<File> {
        let fp = File::open(mountpoint)?;

        if xfs_handle::try_from(&fp)? != self.handle {
            let s = format!(
                "{} {}: {}",
                M_("reopening"),
                mountpoint.display(),
                M_("Stale file handle")
            );
            return Err(Error::new(ErrorKind::Other, s));
        }

        Ok(fp)
    }

    /// Try to reopen the filesystem from which we got the handle.
    pub fn reopen(&self) -> Result<File> {
        // First try the original mountpoint
        let orig_result = self.reopen_from(&self.mountpoint);
        if let Ok(x) = orig_result {
            return Ok(x);
        }

        // Now scan /proc/self/mounts for any other bind mounts of this filesystem
        let entries = MountEntries::try_new()?;
        for mntent in entries.filter(|x| x.fstype == "xfs" && x.fsname == self.fsname) {
            if let Ok(x) = self.reopen_from(&mntent.dir) {
                return Ok(x);
            }
        }

        // Return original error
        orig_result
    }

    /// Report mountpoint in a displayable manner
    pub fn mountpoint(&self) -> String {
        self.mountpoint.display().to_string()
    }

    /// Report xfs device in a displayable manner
    pub fn fsname(&self) -> String {
        self.fsname.clone()
    }

    /// Create a soft handle from an open file descriptor and its mount point
    pub fn try_new(
        fp: &File,
        mountpoint: Arc<PathBuf>,
        fsgeom: xfs_fsop_geom,
    ) -> Result<WeakHandle> {
        let mut entries = MountEntries::try_new()?;
        let fsname = match entries.find(|x| x.fstype == "xfs" && x.dir == *mountpoint) {
            None => {
                let s = format!("{}: {}", mountpoint.display(), M_("Cannot find xfs device"));
                return Err(Error::new(ErrorKind::Other, s));
            }
            Some(mntent) => mntent.fsname,
        };

        Ok(WeakHandle {
            mountpoint,
            fsname,
            handle: xfs_handle::try_from(fp)?,
            has_parent: fsgeom.has_parent(),
        })
    }

    /// Create a new file handle from this one
    pub fn subst(&self, fid: XfsFid) -> xfs_handle {
        xfs_handle {
            ha_fid: xfs_fid {
                fid_ino: fid.ino.into(),
                fid_gen: fid.gen.into(),
                ..self.handle.ha_fid
            },
            ..self.handle
        }
    }

    /// Can this filesystem do parent pointer lookups?
    pub fn can_get_parents(&self) -> bool {
        self.has_parent
    }

    /// Compute the systemd instance unit name for this mountpoint.
    pub fn instance_unit_name(&self, service_template: &str) -> Result<OsString> {
        let output = Command::new("systemd-escape")
            .arg("--template")
            .arg(service_template)
            .arg("--path")
            .arg(self.mountpoint.as_ref())
            .output()?;

        if !output.status.success() {
            return Err(badness!("Could not format systemd instance unit name."));
        }

        // systemd always adds a newline to the end of the output; remove it
        let trunc_out = &output.stdout[0..output.stdout.len() - 1];
        Ok(OsString::from_vec(trunc_out.to_vec()))
    }
}

impl Display for WeakHandle {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        write!(f, "{} {}", self.fsname, self.mountpoint.display())
    }
}

File: healer/weakhandle.c
// SPDX-License-Identifier: GPL-2.0-or-later
/*
 * Copyright (C) 2025-2026 Oracle.  All Rights Reserved.
 * Author: Darrick J. Wong <djwong@kernel.org>
 */
#include "xfs.h"
#include <pthread.h>
#include <stdlib.h>

#include "platform_defs.h"
#include "handle.h"
#include "libfrog/fsgeom.h"
#include "libfrog/workqueue.h"
#include "libfrog/getparents.h"
#include "libfrog/paths.h"
#include "libfrog/systemd.h"
#include "libfrog/statmount.h"
#include "xfs_healer.h"

struct weakhandle {
	/* Synchronizes accesses to mntpoint */
	pthread_rwlock_t	lock;

	/* Owned reference to the user's mountpoint for logging */
	char			*mntpoint;

	/* Shared reference to the getmntent fsname for reconnecting */
	const char		*fsname;

	/* Mount id for faster reconnecting */
	uint64_t		mnt_id;

	/* handle to root dir */
	void			*hanp;
	size_t			hlen;
};

/* Capture a handle for a given filesystem, but don't attach to the fd. */
int
weakhandle_alloc(
	int			fd,
	const char		*mountpoint,
	uint64_t		mnt_id,
	const char		*fsname,
	struct weakhandle	**whp)
{
	struct weakhandle	*wh;
	int			ret;

	*whp = NULL;

	if (fd < 0 || !mountpoint) {
		errno = EINVAL;
		return -1;
	}

	wh = calloc(1, sizeof(struct weakhandle));
	if (!wh)
		return -1;

	wh->mntpoint = strdup(mountpoint);
	if (!wh->mntpoint)
		goto out_wh;
	wh->mnt_id = mnt_id;
	wh->fsname = fsname;

	ret = pthread_rwlock_init(&wh->lock, NULL);
	if (ret) {
		errno = ret;
		goto out_mntpt;
	}

	ret = fd_to_handle(fd, &wh->hanp, &wh->hlen);
	if (ret)
		goto out_rwlock;

	*whp = wh;
	return 0;

out_rwlock:
	pthread_rwlock_destroy(&wh->lock);
out_mntpt:
	free(wh->mntpoint);
out_wh:
	free(wh);
	return -1;
}

static void
try_update_mntpoint(
	struct weakhandle	*wh,
	const char		*path)
{
	char			*s = strdup(path);

	if (!s)
		return;

	pthread_rwlock_wrlock(&wh->lock);
	if (path != wh->mntpoint) {
		free(wh->mntpoint);
		wh->mntpoint = s;
		s = NULL;
	}
	pthread_rwlock_unlock(&wh->lock);

	free(s);
}

/*
 * Reopen a file handle obtained via weak reference, using the given path to a
 * mount point.
 */
static int
weakhandle_reopen_from(
	struct weakhandle	*wh,
	const char		*path,
	int			*fd,
	weakhandle_fd_t		is_acceptable,
	void			*data)
{
	void			*hanp;
	size_t			hlen;
	int			mnt_fd;
	int			ret;

	*fd = -1;

	mnt_fd = open(path, O_RDONLY);
	if (mnt_fd < 0)
		return -1;

	ret = fd_to_handle(mnt_fd, &hanp, &hlen);
	if (ret)
		goto out_mntfd;

	if (hlen != wh->hlen || memcmp(hanp, wh->hanp, hlen)) {
		errno = ESTALE;
		goto out_handle;
	}

	if (is_acceptable && !is_acceptable(mnt_fd, data)) {
		errno = ESTALE;
		goto out_handle;
	}

	free_handle(hanp, hlen);
	*fd = mnt_fd;
	return 0;

out_handle:
	free_handle(hanp, hlen);
out_mntfd:
	close(mnt_fd);
	return -1;
}

/* three paths, a filesystem type string of unknown length, and options */
#define MNTENT_BUFLEN		(4 * PATH_MAX + NAME_MAX)

/* Reopen a file handle obtained via weak reference. */
int
weakhandle_reopen(
	struct weakhandle	*wh,
	int			*fd,
	weakhandle_fd_t		is_acceptable,
	void			*data)
{
	struct mntent		ent;
	const size_t		smbuf_size =
		libfrog_statmount_sizeof(PATH_MAX);
	struct statmount	*smbuf = alloca(smbuf_size);
	FILE			*mtab;
	struct mntent		*mnt;
	char			*buf;
	int			ret;

	/* First try reopening using the original mountpoint */
	pthread_rwlock_rdlock(&wh->lock);
	ret = weakhandle_reopen_from(wh, wh->mntpoint, fd, is_acceptable, data);
	pthread_rwlock_unlock(&wh->lock);
	if (!ret)
		return 0;

	/*
	 * The original mountpoint didn't work, which means the mount might
	 * have been moved.  Look up the mountpoint for the mount id that we
	 * captured earlier, which is a quick lookup if there are many mounts.
	 * Note that @ret is nonzero here.
	 */
	ret = libfrog_statmount(wh->mnt_id, DEFAULT_MOUNTNS_FD,
			STATMOUNT_MNT_POINT, smbuf, smbuf_size);
	if (ret || !(smbuf->mask & STATMOUNT_MNT_POINT))
		goto fallback;
	ret = weakhandle_reopen_from(wh, smbuf->str + smbuf->mnt_point, fd,
			is_acceptable, data);
	if (!ret) {
		try_update_mntpoint(wh, smbuf->str + smbuf->mnt_point);
		return 0;
	}

fallback:
	/*
	 * That didn't work, so now walk /proc/mounts to find a mount with the
	 * same fsname (aka xfs data device path) as when we started.
	 */
	mtab = setmntent(_PATH_PROC_MOUNTS, "r");
	if (!mtab)
		return -1;

	buf = malloc(MNTENT_BUFLEN);
	if (!buf) {
		ret = -1;
		goto out_mtab;
	}

	while ((mnt = getmntent_r(mtab, &ent, buf, MNTENT_BUFLEN)) != NULL) {
		if (strcmp(mnt->mnt_type, "xfs"))
			continue;
		if (strcmp(mnt->mnt_fsname, wh->fsname))
			continue;

		ret = weakhandle_reopen_from(wh, mnt->mnt_dir, fd,
				is_acceptable, data);
		if (!ret) {
			try_update_mntpoint(wh, mnt->mnt_dir);
			break;
		}
	}

	if (*fd < 0) {
		errno = ESTALE;
		ret = -1;
	}

	free(buf);
out_mtab:
	endmntent(mtab);
	return ret;
}

/* Tear down a weak handle */
void
weakhandle_free(
	struct weakhandle	**whp)
{
	struct weakhandle	*wh = *whp;

	if (wh) {
		pthread_rwlock_destroy(&wh->lock);
		free_handle(wh->hanp, wh->hlen);
		free(wh->mntpoint);
		free(wh);
	}

	*whp = NULL;
}

struct bufvec {
	struct weakhandle	*wh;
	char			*buf;
	size_t			len;
};

static int
render_path(
	const char		*donotuse,
	const struct path_list	*path,
	void			*arg)
{
	struct bufvec		*args = arg;
	struct weakhandle	*wh = args->wh;
	int			mntpt_len;
	ssize_t			ret;

	/* Trim trailing slashes from the mountpoint */
	pthread_rwlock_rdlock(&wh->lock);
	mntpt_len = strlen(wh->mntpoint);
	while (mntpt_len > 0 && wh->mntpoint[mntpt_len - 1] == '/')
		mntpt_len--;

	ret = snprintf(args->buf, args->len, "%.*s", mntpt_len, wh->mntpoint);
	pthread_rwlock_unlock(&wh->lock);
	if (ret < 0 || ret >= args->len)
		return 0;

	ret = path_list_to_string(path, args->buf + ret, args->len - ret);
	if (ret < 0)
		return 0;

	/* magic code that means we found one */
	return ECANCELED;
}

/* Render any path to this weakhandle into the specified buffer. */
int
weakhandle_getpath_for(
	struct weakhandle	*wh,
	uint64_t		ino,
	uint32_t		gen,
	char			*path,
	size_t			pathlen)
{
	struct xfs_handle	fakehandle;
	struct bufvec		bv = {
		.wh		= wh,
		.buf		= path,
		.len		= pathlen,
	};
	int			mnt_fd;
	int			ret;

	if (wh->hlen != sizeof(fakehandle)) {
		errno = EINVAL;
		return -1;
	}
	memcpy(&fakehandle, wh->hanp, sizeof(fakehandle));
	fakehandle.ha_fid.fid_ino = ino;
	fakehandle.ha_fid.fid_gen = gen;

	ret = weakhandle_reopen(wh, &mnt_fd, NULL, NULL);
	if (ret)
		return ret;

	/*
	 * In the common case, files only have one parent; and what's the
	 * chance that we'll need to walk past the second parent to find *one*
	 * path that goes to the rootdir?  With a max filename length of 255
	 * bytes, we pick 600 for the buffer size.
	 */
	ret = handle_walk_paths_fd(NULL, mnt_fd, &fakehandle,
			sizeof(fakehandle), 600, render_path, &bv);
	switch (ret) {
	case ECANCELED:
		/* found a path */
		ret = 0;
		break;
	default:
		/* didn't find one */
		errno = ENOENT;
		ret = -1;
		break;
	}

	close(mnt_fd);
	return ret;
}

/* Compute the systemd instance unit name for this mountpoint. */
int
weakhandle_instance_unit_name(
	struct weakhandle	*wh,
	const char		*template,
	char			*unitname,
	size_t			unitnamelen)
{
	int			ret;

	pthread_rwlock_rdlock(&wh->lock);
	ret = systemd_path_instance_unit_name(template, wh->mntpoint,
			unitname, unitnamelen);
	pthread_rwlock_unlock(&wh->lock);
	return ret;
}
